doc/tutorial/concepts/index.md

```diff
@@ -0,0 +1,214 @@
+
+\tutorial_head{Moderate, 60 min, \ref setup\, \ref tutorial_alphabets,}
+
+This tutorial teaches the very basics of working with concepts. For more background and information on how to implement
+your own concepts, we recommend:
+  * A well-readable [paper](http://www.stroustrup.com/good_concepts.pdf) with motivation and historical background.
+  * The (rather dense) [documentation on cppreference](https://en.cppreference.com/w/cpp/language/constraints).
+
+
+# Constraints
+
+## Introduction
+
+In C++20 a typical pre-C++20 template is said to be *unconstrained*, i.e. it can be instantiated with any type:
+
+```cpp
+template <typename t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+Of course this will not actually compile for all types, because the function template **implicitly requires** that
+types provide a `+` operator.
+*Constraints* are a way of making requirements of template arguments **explicit**.
+Constraints can be formulated ad-hoc, but this tutorial only covers *concepts*.
+Concepts are a set of constraints with a given name.
+Let's assume there is a concept called `Addable` that requires the existance of a `+` operator (as previously mentioned
+the syntax for defining concepts is not covered here).
+The following snippet demonstrates how we can constrain our function template, i.e. make the template immediately
+reject any types that don't satisfy the requirement:
+
+```cpp
+template <Addable t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+The only difference is that we have replaced `typename` with `Addable`. The standard library provides a set of
+[predefined concepts](https://en.cppreference.com/w/cpp/concepts), for our example above, the std::Integral concept
+could have been used.
+
+## Syntax variants
+
+Different constraints can be applied to different template parameters and a single template parameter can be constrained
+by multiple concepts.
+Depending on the complexity of your constraint statements, three different syntaxes are available to enforce
+constraints; all of the following are equivalent.
+
+(1) The "verbose syntax", especially useful when enforcing multiple constraints:
+
+```cpp
+template <typename t1, typename t2>
+    requires std::Integral<t1> && std::Integral<t2> // && MyOtherConcept<t1>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(2) The "intermediate syntax":
+```cpp
+template <std::Integral t1, std::Integral t2>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(3) The "terse syntax":
+```cpp
+auto add(std::Integral auto const v1, std::Integral const v2)
+{
+    return v1 + v2;
+}
+```
+
+\attention The terse syntax in this form is not yet available in GCC7, GCC8 and GCC9.
```

Comment:
Please add a small paragraph stating how we use it in SeqAn.
Maybe also add an example for a the intermediate syntax with a constrained that uses at least two template arguments.

Reply:
> Please add a small paragraph stating how we use it in SeqAn.

I am not sure if this soo relevant for people that just use our interfaces to write apps (and don't read our code). There is a section further down on how concepts are documented in our docs.

> Maybe also add an example for a the intermediate syntax with a constrained that uses at least two template arguments.

Hm, I'd like to keep this short and really focus the necessary bits to get along and write your app. The syntax you are referring to is cool, but rarely used, especially in apps I think.

Reply:
I would say the  "How we use it in seqan3" is maybe rather for developers and contributors than for users of the library

Reply:
But the terse syntax is not used either, since it is not yet available. And for me it is less the fact that users write this code but need to read the code and the documentation. So for them it is important to understand of how we use it.  But it is true that there is a separate section with regard to doxygen and this would be the better place to provide such information. I'll check this section again